src/main/java/org/eclipse/microprofile/starter/rest/APIService.java

```diff
@@ -78,18 +82,38 @@
 
     @PostConstruct
     public void init() {
+        // Keys are MP versions and values are servers and specs
         mpvToOptions =
                 new HashMap<MicroProfileVersion, MPOptionsAvailable>(MicroProfileVersion.values().length) {
                     {
-                        Stream.of(MicroProfileVersion.values()).forEach(mpv -> put(mpv, new MPOptionsAvailable(
-                                Stream.of(SupportedServer.values())
-                                        .filter(v -> v.getMpVersions().contains(mpv)).collect(Collectors.toList()),
-                                Stream.of(MicroprofileSpec.values())
-                                        .filter(v -> v.getMpVersions().contains(mpv)).collect(Collectors.toList()))));
+                        Stream.of(MicroProfileVersion.values()).filter(mpv -> mpv != MicroProfileVersion.NONE)
+                                .forEach(mpv -> put(mpv, new MPOptionsAvailable(
+                                        Stream.of(SupportedServer.values())
+                                                .filter(v -> v.getMpVersions().contains(mpv)).collect(Collectors.toList()),
+                                        Stream.of(MicroprofileSpec.values())
+                                                .filter(v -> v.getMpVersions().contains(mpv)).collect(Collectors.toList()))));
 
                     }
 
                 };
+        mpvToOptionsEtag = new EntityTag(Integer.toHexString(31 * version.getGit().hashCode() + mpvToOptions.hashCode()));
+        // Keys are servers and values are MP versions and specs
+        serversToOptions =
```

Comment:
Same remark as for mpvToOptions -> unreadble!

This is equivalent (this does not include the random order, see other comment)

        serversToOptions = new HashMap<>();
        for (SupportedServer supportedServer : SupportedServer.values()) {
            Map<MicroProfileVersion, List<MicroprofileSpec>> serverOption = new HashMap<>();
            for (MicroProfileVersion mpVersion : supportedServer.getMpVersions()) {
                List<MicroprofileSpec> specList = new ArrayList<>();
                for (MicroprofileSpec spec : MicroprofileSpec.values()) {
                    if (spec.getMpVersions().contains(mpVersion)) {
                        specList.add(spec);
                    }
                }
                serverOption.put(mpVersion, specList);

            }
            serversToOptions.put(supportedServer, serverOption);
        }

or with a little bit of Stream ...

        serversToOptions = new HashMap<>();
        for (SupportedServer supportedServer : SupportedServer.values()) {
            Map<MicroProfileVersion, List<MicroprofileSpec>> serverOption = new HashMap<>();
            for (MicroProfileVersion mpVersion : supportedServer.getMpVersions()) {
                List<MicroprofileSpec> specList =
                        Arrays.stream(MicroprofileSpec.values())
                                .filter(ms -> ms.getMpVersions().contains(mpVersion))
                                .collect(Collectors.toList());
                serverOption.put(mpVersion, specList);

            }
            serversToOptions.put(supportedServer, serverOption);
        }

Reply:
@rdebusscher Refactored.